### DISCUSSION TITLE
3D/4D porosity networks

### DISCUSSION THREAD

**Discussion by DVangelis** [PMEAL/OpenPNM#2527] [Q&A]
Hi I am quite new to porsepy and openpnm. I want to see how porous and trhoats are changing in the three principle axes X Y Z. 
I have a binary image volume and I have managed to export the network of this volume with using the snow algorithm. I have also managed to view it with paraview. My question is if it is possible with openpnm to get information for throats and porous in the three different main orientations  X Y Z and plot it in a histogram for example. When I will achieve this I would like to further plot the change of this 3d structure with time any recommendations of how I could do that it would be higgle appreciated. Thanks in advance .

**Comment by jgostick** [comment]
If you have the network in OpenPNM already then you can computer the vector of each throat by looking at the coordinates of the pores on each end.  For instance you can do:

``` python
crds_1 = pn['pore.coords'][pn['throat.conns'][:, 0]]
crds_2 = pn['pore.coords'][pn['throat.conns'][:, 1]]

# The above can be done with one line if you use a few tricks:
crds1, crds2 = pn.coords[pn.conns.T]

```

You then have the [x, y, z] coords of the pores on each end, so could compute the vector or look at the length.  The trick would be that each image you process will have different labels for the pores, so tracking the movement of an exact set of pores will be hard.  You can do overalls stats and histograms though.

**Reply by DVangelis** [reply]
Thanks for the reply .
I tried your code .
So the crds_1 returns one numpy array containing the coordinates of the pores in the direction 0 that correspond to one of the main axis ? Am I correct ?
And how I would compute the vector or lenght of pores . Just be using crds_1[0] and crds_1[1] for example ?
Thanks in advance.

**Reply by jgostick** [reply]
crds_0 is the [x, y, z] coordinates of the pore on one end of each throat, and crds_1 is the coordinates of the pore on the other end.  Since a pore can have many throats, it's coordinates may appear many times.  Also, the defintion of which pore is on one end or the other is aribtrary, so a given pore's coordinates may appear in crds_0 or crds_1.

To get vector or length you just use pythagorean thereom and euclidean geometry to compute the distances.  

For the length you would use:

``` python

L = np.sqrt(np.sum((crds_0 - crds_1)**2, axis=1))

```

To get the components of the vector use:

``` python
x, y, z = (crds_0 - crds_1).T
```

Then of course you can normalize the ``x, y, z`` values by ``L`` to get the components of the unit vector.
